File: java/src/main/java/Oberon.java
import com.google.protobuf.InvalidProtocolBufferException;
import trinsic.okapi.security.*;

public class Oberon extends OkapiNative {
    public static CreateOberonKeyResponse createKey(CreateOberonKeyRequest request) throws DidException, InvalidProtocolBufferException {
        OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
        OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
        ExternError errBuffer = new ExternError();
        getNativeLibrary().oberon_create_key(requestBuffer, responseBuffer, errBuffer);
        errBuffer.RaiseError();
        return CreateOberonKeyResponse.parseFrom(bufferToByteArray(responseBuffer));
    }

    public static CreateOberonTokenResponse createToken(CreateOberonTokenRequest request) throws DidException, InvalidProtocolBufferException {
        OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
        OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
        ExternError errBuffer = new ExternError();
        getNativeLibrary().oberon_create_token(requestBuffer, responseBuffer, errBuffer);
        errBuffer.RaiseError();
        return CreateOberonTokenResponse.parseFrom(bufferToByteArray(responseBuffer));
    }

    public static BlindOberonTokenResponse blindToken(BlindOberonTokenRequest request) throws DidException, InvalidProtocolBufferException {
        OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
        OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
        ExternError errBuffer = new ExternError();
        getNativeLibrary().oberon_blind_token(requestBuffer, responseBuffer, errBuffer);
        errBuffer.RaiseError();
        return BlindOberonTokenResponse.parseFrom(bufferToByteArray(responseBuffer));
    }

    public static UnBlindOberonTokenResponse unBlindToken(UnBlindOberonTokenRequest request) throws DidException, InvalidProtocolBufferException {
        OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
        OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
        ExternError errBuffer = new ExternError();
        getNativeLibrary().oberon_unblind_token(requestBuffer, responseBuffer, errBuffer);
        errBuffer.RaiseError();
        return UnBlindOberonTokenResponse.parseFrom(bufferToByteArray(responseBuffer));
    }

    public static CreateOberonProofResponse createProof(CreateOberonProofRequest request) throws DidException, InvalidProtocolBufferException {
        OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
        OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
        ExternError errBuffer = new ExternError();
        getNativeLibrary().oberon_create_proof(requestBuffer, responseBuffer, errBuffer);
        errBuffer.RaiseError();
        return CreateOberonProofResponse.parseFrom(bufferToByteArray(responseBuffer));
    }

    public static VerifyOberonProofResponse verifyProof(VerifyOberonProofRequest request) throws DidException, InvalidProtocolBufferException {
        OkapiByteBuffer.ByValue requestBuffer = messageToBuffer(request);
        OkapiByteBuffer responseBuffer = new OkapiByteBuffer();
        ExternError errBuffer = new ExternError();
        getNativeLibrary().oberon_verify_proof(requestBuffer, responseBuffer, errBuffer);
        errBuffer.RaiseError();
        return VerifyOberonProofResponse.parseFrom(bufferToByteArray(responseBuffer));
    }
}

File: ruby/lib/oberon.rb
# frozen_string_literal: true

# Oberon module
module Oberon
  def self.create_key(request)
    Okapi.verify_type(request, Security_V1::CreateOberonKeyRequest)
    Okapi.ffi_call('oberon_create_key', request, Security_V1::CreateOberonKeyResponse)
  end

  def self.create_token(request)
    Okapi.verify_type(request, Security_V1::CreateOberonTokenRequest)
    Okapi.ffi_call('oberon_create_token', request, Security_V1::CreateOberonTokenResponse)
  end

  def self.blind_token(request)
    Okapi.verify_type(request, Security_V1::BlindOberonTokenRequest)
    Okapi.ffi_call('oberon_blind_token', request, Security_V1::BlindOberonTokenResponse)
  end

  def self.unblind_token(request)
    Okapi.verify_type(request, Security_V1::UnBlindOberonTokenRequest)
    Okapi.ffi_call('oberon_unblind_token', request, Security_V1::UnBlindOberonTokenResponse)
  end

  def self.create_proof(request)
    Okapi.verify_type(request, Security_V1::CreateOberonProofRequest)
    Okapi.ffi_call('oberon_create_proof', request, Security_V1::CreateOberonProofResponse)
  end

  def self.verify_proof(request)
    Okapi.verify_type(request, Security_V1::VerifyOberonProofRequest)
    Okapi.ffi_call('oberon_verify_proof', request, Security_V1::VerifyOberonProofResponse)
  end
end
